src/foam/nanos/pm/FoldReducePMLogger.java

```diff
@@ -6,116 +6,154 @@
+    StringBuilder trace = new StringBuilder("// Captured at: ");
+    trace.append(new Date());
+    trace.append(System.lineSeparator());
 
-  public void capture(PMInfo pi) {
-    StringBuilder trace = new StringBuilder();
     for ( StackTraceElement j : Thread.currentThread().getStackTrace() ) {
       trace.append(j.toString());
-      trace.append(System.getProperty("line.separator"));
+      trace.append(System.lineSeparator());
     }
 
-    pi.setCapture(false);
-    pi.setCaptureTrace(trace.toString());
+    pmi.setCapture(false);
+    pmi.setCaptureTrace(trace.toString());
+    updateCaptureEnabledList(pmi);
   }
 
+  /** Manage keys for maintaining capture flag between state and LocalState **/
+  public synchronized void updateCaptureEnabledList(PMInfo pmi) {
+    if ( pmi.getCapture() ) {
+      captureEnabledList_.add(pmi);
+    } else {
+      captureEnabledList_.remove(pmi);
+    }
+  }
 
   /** Template method to Merge two states. **/
   public Object reduce(Object state1, Object state2) {
     Map<String,PMInfo> m1 = (Map<String,PMInfo>) state1;
     Map<String,PMInfo> m2 = (Map<String,PMInfo>) state2;
 
-    for ( PMInfo pi2 : m2.values() ) {
-      String key = pi2.getKey() + ":" + pi2.getName();
-      PMInfo pi1 = m1.get(key);
+    for ( PMInfo pmi2 : m2.values() ) {
+      String key = pmi2.getKey() + ":" + pmi2.getName();
+      PMInfo pmi1 = m1.get(key);
 
-      if ( pi1 == null ) {
-        m1.put(key, pi2);
+      if ( pmi1 == null ) {
+        m1.put(key, pmi2);
       } else {
-        pi1.reduce(pi2);
+        pmi1.reduce(pmi2);
       }
     }
 
     return state1;
   }
 
-
   @Override
   public void log(PM pm) {
     if (
       ! pm.getKey().equals("foam.dao.PMDAO") &&
       ! pm.getKey().equals("foam.dao.PipelinePMDAO") &&
       ! pm.getKey().equals("foam.nanos.auth.PMAuthService")
     ) {
-      if ( pm.getKey().indexOf("PM")  != -1 ) return;
-      if ( pm.getName().indexOf("PM") != -1 ) return;
-      if ( pm.getKey().indexOf("pm")  != -1 ) return;
-      if ( pm.getName().indexOf("pm") != -1 ) return;
-      if ( pm.getName().indexOf("LogMessage") != -1 ) return;
+      if ( pm.getKey().contains("PM") ) return;
+      if ( pm.getName().contains("PM") ) return;
+      if ( pm.getKey().contains("pm") ) return;
+      if ( pm.getName().contains("pm") ) return;
+      if ( pm.getName().contains("LogMessage") ) return;
     }
 
     this.fold(pm);
   }
 
-
   public synchronized DAO asDAO() {
     return new ProxyDAO() {
       public DAO getDelegate() {
         synchronized ( FoldReducePMLogger.this ) {
-          Map  m   = (Map) getState();
-          MDAO dao = new MDAO(foam.nanos.pm.PMInfo.getOwnClassInfo());
+          DAO dao = new PMInfoMDAO();
+          Map<String, PMInfo>  m   = (Map<String, PMInfo>) getState();
 
-          for ( Object pi : m.values() ) dao.put((PMInfo) pi);
+          for ( PMInfo pi : m.values() ) {
+            pi.setCapture(captureEnabledList_.contains(pi));
```

Comment:
Is this line necessary?

Reply:
yes this is necessary to render correctly in the ui when changing the state of the capture flag